File: datawise-frontend/electron/workspace-preferences.ts
import {existsSync, mkdirSync} from 'node:fs'
import {join, normalize} from 'node:path'
import {app} from 'electron'
import {readDesktopPreferences, writeDesktopPreferences} from './desktop-preferences'

export const MAX_RECENT_WORKSPACES = 8

export interface WorkspaceListEntry {
    path: string
    active: boolean
    isDefault: boolean
}

export type CreateWorkspaceResult =
    | {ok: true; path: string}
    | {ok: false; error: 'invalid' | 'exists'}

export function sanitizeWorkspaceFolderName(name: string): string | null {
    const trimmed = name.trim()
    if (!trimmed) return null
    if (/[<>:"|?*\x00/\\]/.test(trimmed)) return null
    if (trimmed === '.' || trimmed === '..') return null
    return trimmed
}

export function workspaceDirectoryHasConfig(dir: string): boolean {
    return existsSync(join(dir, 'users.json'))
        || existsSync(join(dir, 'connections.xml'))
        || existsSync(join(dir, 'sessions.json'))
}

export function resolveNewWorkspaceParent(): string {
    const parent = join(app.getPath('documents'), 'DataWise', 'workspaces')
    mkdirSync(parent, {recursive: true})
    return parent
}

export function prepareNewWorkspaceDirectory(name: string): CreateWorkspaceResult {
    const safe = sanitizeWorkspaceFolderName(name)
    if (!safe) return {ok: false, error: 'invalid'}

    const target = join(resolveNewWorkspaceParent(), safe)
    if (existsSync(target) && workspaceDirectoryHasConfig(target)) {
        return {ok: false, error: 'exists'}
    }

    mkdirSync(target, {recursive: true})
    return {ok: true, path: normalize(target)}
}

export function normalizeWorkspacePath(path: string): string {
    const trimmed = path.trim()
    if (!trimmed) return ''
    return normalize(trimmed)
}

export function touchRecentWorkspace(resolvedPath: string): string[] {
    const key = normalizeWorkspacePath(resolvedPath)
    if (!key) return readDesktopPreferences().recentWorkspaces ?? []

    const prev = readDesktopPreferences().recentWorkspaces ?? []
    const next = [key, ...prev.filter((item) => normalizeWorkspacePath(item) !== key)]
        .slice(0, MAX_RECENT_WORKSPACES)
    writeDesktopPreferences({recentWorkspaces: next})
    return next
}

export function removeRecentWorkspace(resolvedPath: string): string[] {
    const key = normalizeWorkspacePath(resolvedPath)
    const prev = readDesktopPreferences().recentWorkspaces ?? []
    const next = prev.filter((item) => normalizeWorkspacePath(item) !== key)
    writeDesktopPreferences({recentWorkspaces: next})
    return next
}

export function buildWorkspaceList(activePath: string, defaultPath: string): WorkspaceListEntry[] {
    const activeNorm = normalizeWorkspacePath(activePath)
    const defaultNorm = normalizeWorkspacePath(defaultPath)
    const recent = readDesktopPreferences().recentWorkspaces ?? []

    const ordered: string[] = []
    const seen = new Set<string>()

    const push = (path: string) => {
        const normalized = normalizeWorkspacePath(path)
        if (!normalized || seen.has(normalized)) return
        seen.add(normalized)
        ordered.push(normalized)
    }

    push(activePath)
    for (const item of recent) push(item)
    push(defaultPath)

    const withoutActive = ordered.filter((path) => path !== activeNorm)
    const finalPaths = activeNorm ? [activeNorm, ...withoutActive] : withoutActive

    return finalPaths.map((path) => ({
        path,
        active: path === activeNorm,
        isDefault: path === defaultNorm,
    }))
}

File: datawise-desktop/src/main/java/org/apache/datawise/desktop/prefs/WorkspacePreferences.java
package org.apache.datawise.desktop.prefs;

import com.google.gson.JsonObject;
import org.apache.datawise.desktop.DesktopPaths;

import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Set;

public final class WorkspacePreferences {
    private WorkspacePreferences() {
    }

    public static String sanitizeFolderName(String name) {
        String trimmed = name == null ? "" : name.trim();
        if (trimmed.isEmpty()) {
            return null;
        }
        if (trimmed.matches(".*[<>:\"|?*\\x00/\\\\].*") || ".".equals(trimmed) || "..".equals(trimmed)) {
            return null;
        }
        return trimmed;
    }

    public static boolean hasConfig(Path dir) {
        return Files.isRegularFile(dir.resolve("users.json"))
                || Files.isRegularFile(dir.resolve("connections.xml"))
                || Files.isRegularFile(dir.resolve("sessions.json"))
                || Files.isRegularFile(dir.resolve("tenants").resolve("index.json"))
                || Files.isRegularFile(dir.resolve("tenants").resolve("default").resolve("connections.xml"));
    }

    public static Path resolveNewWorkspaceParent() {
        Path parent = DesktopPaths.documentsWorkspacesParent();
        try {
            Files.createDirectories(parent);
        } catch (Exception e) {
            throw new IllegalStateException("Cannot create workspace parent: " + parent, e);
        }
        return parent;
    }

    public static JsonObject prepareNewWorkspace(String name) {
        String safe = sanitizeFolderName(name);
        JsonObject result = new JsonObject();
        if (safe == null) {
            result.addProperty("ok", false);
            result.addProperty("error", "invalid");
            return result;
        }
        Path target = resolveNewWorkspaceParent().resolve(safe);
        if (Files.exists(target) && hasConfig(target)) {
            result.addProperty("ok", false);
            result.addProperty("error", "exists");
            return result;
        }
        try {
            Files.createDirectories(target);
        } catch (Exception e) {
            result.addProperty("ok", false);
            result.addProperty("error", "invalid");
            return result;
        }
        result.addProperty("ok", true);
        result.addProperty("path", target.normalize().toString());
        return result;
    }

    public static List<JsonObject> buildWorkspaceList(String activePath, String defaultPath) {
        String activeNorm = Path.of(activePath).normalize().toString();
        String defaultNorm = Path.of(defaultPath).normalize().toString();
        Set<String> ordered = new LinkedHashSet<>();
        if (!activeNorm.isBlank()) {
            ordered.add(activeNorm);
        }
        for (String item : DesktopPreferencesStore.recentWorkspaces()) {
            ordered.add(Path.of(item).normalize().toString());
        }
        if (!defaultNorm.isBlank()) {
            ordered.add(defaultNorm);
        }
        List<JsonObject> out = new ArrayList<>();
        for (String path : ordered) {
            JsonObject entry = new JsonObject();
            entry.addProperty("path", path);
            entry.addProperty("active", path.equals(activeNorm));
            entry.addProperty("isDefault", path.equals(defaultNorm));
            out.add(entry);
        }
        return out;
    }
}
